mm2src/coins/tendermint/tendermint_tx_history_v2.rs

```diff
@@ -0,0 +1,754 @@
+    fn history_wallet_id(&self) -> WalletId { WalletId::new(self.ticker().replace('-', "_")) }
+
+    async fn get_tx_history_filters(
+        &self,
+        _target: MyTxHistoryTarget,
+    ) -> MmResult<GetTxHistoryFilters, MyTxHistoryErrorV2> {
+        Ok(GetTxHistoryFilters::for_address(
+            self.platform_coin.account_id.to_string(),
+        ))
+    }
+}
+
+struct TendermintTxHistoryCtx<Coin: TendermintTxHistoryOps, Storage: TxHistoryStorage> {
+    coin: Coin,
+    storage: Storage,
+    #[allow(dead_code)]
+    metrics: MetricsArc,
+    balances: AllBalancesResult,
+}
+
+struct TendermintInit<Coin, Storage> {
+    phantom: std::marker::PhantomData<(Coin, Storage)>,
+}
+
+impl<Coin, Storage> TendermintInit<Coin, Storage> {
+    fn new() -> Self {
+        TendermintInit {
+            phantom: Default::default(),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum StopReason {
+    StorageError(String),
+    UnknownError(String),
+    RpcClient(String),
+    Marshaling(String),
+}
+
+struct Stopped<Coin, Storage> {
+    phantom: std::marker::PhantomData<(Coin, Storage)>,
+    stop_reason: StopReason,
+}
+
+impl<Coin, Storage> Stopped<Coin, Storage> {
+    fn storage_error<E>(e: E) -> Self
+    where
+        E: std::fmt::Debug,
+    {
+        Stopped {
+            phantom: Default::default(),
+            stop_reason: StopReason::StorageError(format!("{:?}", e)),
+        }
+    }
+
+    fn unknown(e: String) -> Self {
+        Stopped {
+            phantom: Default::default(),
+            stop_reason: StopReason::UnknownError(e),
+        }
+    }
+}
+
+struct WaitForHistoryUpdateTrigger<Coin, Storage> {
+    phantom: std::marker::PhantomData<(Coin, Storage)>,
+}
+
+impl<Coin, Storage> WaitForHistoryUpdateTrigger<Coin, Storage> {
+    fn new() -> Self {
+        WaitForHistoryUpdateTrigger {
+            phantom: Default::default(),
+        }
+    }
+}
+
+struct FetchingTransactionsData<Coin, Storage> {
+    /// The list of addresses for those we have requested [`UpdatingUnconfirmedTxes::all_tx_ids_with_height`] TX hashses
+    /// at the `FetchingTxHashes` state.
+    address: String,
+    /// Denom - Ticker
+    active_assets: HashMap<String, String>,
+    phantom: std::marker::PhantomData<(Coin, Storage)>,
+}
+
+impl<Coin, Storage> FetchingTransactionsData<Coin, Storage> {
+    fn new(address: String, active_assets: HashMap<String, String>) -> Self {
+        FetchingTransactionsData {
+            address,
+            active_assets,
+            phantom: Default::default(),
+        }
+    }
+}
+
+impl<Coin, Storage> TransitionFrom<TendermintInit<Coin, Storage>> for Stopped<Coin, Storage> {}
+impl<Coin, Storage> TransitionFrom<TendermintInit<Coin, Storage>> for FetchingTransactionsData<Coin, Storage> {}
+impl<Coin, Storage> TransitionFrom<WaitForHistoryUpdateTrigger<Coin, Storage>> for Stopped<Coin, Storage> {}
+impl<Coin, Storage> TransitionFrom<FetchingTransactionsData<Coin, Storage>> for Stopped<Coin, Storage> {}
```

Comment:
@sergeyboyko0791 @ozkanonur does this make sense in this case?

```rust
macro_rules! transition_map {
    ($from: ident<$coin: ident, $storage: ident> => $to: ident) => {
        impl<$coin, $storage> TransitionFrom<$from<$coin, $storage>>
            for $to<$coin, $storage> {}
    };
    ($from: ident<$coin: ident, $storage: ident> => [$( $to:ident ),*] ) => {
        $(transition_map!($from<$coin, $storage> => $to);)*
    };
}

transition_map!(TendermintInit<Coin, Storage> => [Stopped, FetchingTransactionsData]);
transition_map!(WaitForHistoryUpdateTrigger<Coin, Storage> => [Stopped, FetchingTransactionsData]);
transition_map!(FetchingTransactionsData<Coin, Storage> => [Stopped, WaitForHistoryUpdateTrigger]);
```

Reply:
@borngraced thank you for the suggestion!
This works for states with 2 generic type params. But I'd like to add a common solution.
For example:
```rust
macro_rules! transition_map {
    ($from: ident $(< $( $gt:tt ),+ >)? => $to: ident) => {
        impl $(< $( $gt ),+ >)? TransitionFrom<$from $(< $( $gt ),+ >)? >
            for $to $(< $( $gt ),+ >)? {}
    };
    ($from: ident $(< $( $gt:tt ),+ >)? => [$( $to:ident ),*] ) => {
        $(transition_map!($from $(< $( $gt ),+ >)? => $to);)*
    };
}

transition_map!(TendermintInit<Coin, Storage> => [Stopped, FetchingTransactionsData]);
transition_map!(WaitForHistoryUpdateTrigger<Coin, Storage> => [Stopped, FetchingTransactionsData]);
transition_map!(FetchingTransactionsData<Coin, Storage> => [Stopped, WaitForHistoryUpdateTrigger]);
```
But it doesn't work, because the generic type params repeat multiple times:
```bash
error: meta-variable `gt` repeats 1 time, but `to` repeats 2 times
   --> src/main.rs:139:10
    |
139 |         $(transition_map!($from $(< $( $gt ),+ >)? => $to);)*
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

Reply:
make sense and yes it's very complicated! 😄

Reply:
Seems more chaotic than the current implementation :)